### DISCUSSION TITLE
Defaults are baked in during API generation

### DISCUSSION THREAD

**Discussion by crummy** [RobinTail/express-zod-api#1706] [Q&A] [answered]
I'm setting up the openAPI generator, and swagger hosting (really nice, thanks!) I've noticed one problem, however.

Given a zod field like this:
```
endDate: z
  .string()
  .datetime()
  .default(() => DateTime.utc().toISO())
  .describe(
    "An ISO8601 UTC string indicating the latest date before which you want to limit results.",
  ),
```
I get documentation like this:
<img width="829" alt="image" src="https://github.com/RobinTail/express-zod-api/assets/690626/d569adbf-7e5c-490c-8cbe-ca7e646c7591">
This looks fine except that the time does not change. So, if I generate the docs now, and wait a few days, the default is actually wrong.

I tried this:
```
endDate: withMeta(z
  .string()
  .datetime()
  .default(() => DateTime.utc().toISO())
  .describe(
    "An ISO8601 UTC string indicating the latest date before which you want to limit results.",
  )).default("Today"),
```
... but that just overrode the actual default.

Any other solutions to this?

**Comment by RobinTail** [comment]
Hello @crummy ,

after generating OpenAPI specification it becomes a static data.
The `default` property of the `SchemaObject` inherits from JSON Schema and it's a static value as well, not an expression.
So it can not be calculated at the moment when you open the generated documentation.

This is a known limitation, see: https://github.com/OAI/OpenAPI-Specification/issues/554

There are suggestions to involve some custom properties, but their actual rendering should be supported by an external tool or plugin that I'm not aware of. Consider asking OAS and Swagger community.

In case you'll find out more about it, I'd be glad to adopt such solution into the generator.

**Reply by crummy** [reply]
That makes sense. Would it be possible to add a field to `withMeta` that would override the default for purpose of documentation, or is that not straightforward? I think the doc_default field shares a similar purpose: https://github.com/marshmallow-code/apispec/pull/247/files

**Reply by RobinTail** [reply]
I'm not sure I understand how another field in the schema could make the default value change on another day after you had generated it, @crummy ?

**Reply by crummy** [reply]
I don't need to show the latest actual value (e.g. `2024-04...`) - instead I'd like to replace the text in the documentation with something like `Today's date, minus 30 days`.

**Reply by RobinTail** [reply]
Ah, got it, @crummy 
Thank you for the clarification.

**Reply by RobinTail** [reply]
What could be a good name for the method that changes the `default` in documentation only, @crummy ?

**Reply by RobinTail** [reply]
I'm thinking about `.describeDefault()` and considering making it only available when `ZodDefault` is wrapped into `withMeta()`, @crummy 
How do you feel about that?

**Reply by crummy** [reply]
That's perfect, I think that will be an ideal solution.

**Reply by RobinTail** [reply]
Dear @crummy ,

here is the feature PR: https://github.com/RobinTail/express-zod-api/pull/1709
And I released the version `18.4.0-beta1` for you.
Could you please install it, test it and let me know if it works as expected?

**Reply by crummy** [reply]
Yes, that works great! 

<img width="858" alt="image" src="https://github.com/RobinTail/express-zod-api/assets/690626/f565cbf3-19b0-4d01-8cab-f6c3836e6dcd">

**Reply by RobinTail** [reply]
I'm glad to read it, @crummy . Thank you for letting me know.
I will probably make a couple adjustments before release.
Still considering a better naming having less confusion.
I will inform you.

**Reply by RobinTail** [reply, marked as answer]
Dear @crummy ,

version 18.4.0 released 🚀 
The method you need is called `.label()` now.
